kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/LogSegment.java

```diff
@@ -186,6 +191,69 @@ public LogSegment(
                         Comparator.comparing((FileStatus a) -> new Path(a.getPath()).getName())
                             .reversed())
                     .collect(Collectors.toList()));
+
+    this.compactionsReversed =
+        new Lazy<>(
+            () ->
+                compactions.stream()
+                    .sorted(
+                        Comparator.comparing((FileStatus a) -> new Path(a.getPath()).getName())
+                            .reversed())
+                    .collect(Collectors.toList()));
+
+    this.allFilesWithCompactionsReversed = new Lazy<>(() -> loadAllFilesWithCompactionsReversed());
+  }
+
+  private List<FileStatus> loadAllFilesWithCompactionsReversed() {
+    if (compactions.isEmpty()) {
+      return allFilesReversed.get();
+    } else {
+      final Iterator<FileStatus> deltaIt = allFilesReversed.get().iterator();
+      final Iterator<FileStatus> compactionIt = compactionsReversed.get().iterator();
```

Comment:
Since these are lexicographically sorted, does it make sense to process them as an iterator?

```
0010.0011.json
0010.0099.json
0014.0016.json
0015.0017.json
0015.0019.json
0016.0018.json
```

We should really just take `0010.0099.json` and ignore all the others, right?

Reply:
Or do we just want to implement a greedy algorithm that reads the x.y.json in order as we can?

Reply:
I've added a helper that uses a greedy algorithm and is (hopefully) easier to understand than this was.